tianshou/data/utils/converter.py

```diff
@@ -80,3 +82,85 @@ def to_torch_as(
     """
     assert isinstance(y, torch.Tensor)
     return to_torch(x, dtype=y.dtype, device=y.device)
+
+
+# Note: object is used as a proxy for objects that can be pickled
+# Note: mypy does not support cyclic definition currently
+Hdf5ConvertibleValues = Union[  # type: ignore
+    int, float, Batch, np.ndarray, torch.Tensor, object,
+    'Hdf5ConvertibleType',  # type: ignore
+]
+
+Hdf5ConvertibleType = Dict[str, Hdf5ConvertibleValues]  # type: ignore
+
+
+def to_hdf5(x: Hdf5ConvertibleType, y: h5py.Group) -> None:
+    """Copy object into HDF5 group."""
+
+    def to_hdf5_via_pickle(x: object, y: h5py.Group, key: str) -> None:
+        """Pickle, convert to numpy array and write to HDF5 dataset."""
+        data = np.frombuffer(pickle.dumps(x), dtype=np.byte)
+        y.create_dataset(key, data=data)
+
+    for k, v in x.items():
+        if isinstance(v, (Batch, dict)):
+            # dicts and batches are both represented by groups
+            subgrp = y.create_group(k)
+            if isinstance(v, Batch):
+                subgrp_data = v.__getstate__()
+            else:
+                subgrp_data = v
+            subgrp.attrs["__data_type__"] = v.__class__.__name__
+            to_hdf5(subgrp_data, subgrp)
+        elif isinstance(v, (np.ndarray, torch.Tensor)):
+            try:
+                # NumPy arrays and PyTorch tensors are written to datasets
+                y.create_dataset(k, data=to_numpy(v))
+                y[k].attrs["__data_type__"] = v.__class__.__name__
+            except TypeError:
+                # if data type is not supported by HDF5 fall back to pickle
+                try:
+                    to_hdf5_via_pickle(v, y, k)
+                except Exception as e:
+                    raise NotImplementedError(
+                        f"Attempted to pickle {v.__class__.__name__} with key"
```

Comment:
This exception is not very clear in this case.

Reply:
Maybe a `NotImplementedError` isn't the right choice. Would you prefer a `RuntimeError` with a message like "Fallback to pickle failed for {v.__class__.__name__} with key '{k}' which cannot be written to a HDF5 dataset directly."`?

Reply:
Yes `RuntimeError` is better. And instead of mentioning 'key', I would clarify that it is because the datatype of the array is not supported, as you explained to be previously.

Reply:
Makes sense. I have replaced the `NotImplementedError` with a `RuntimeError` and removed the part of the message mentioning the key.